bridges/snowbridge/primitives/outbound-queue/src/v2/converter/convert.rs

```diff
@@ -0,0 +1,270 @@
+		Ok(fee_amount)
+	}
+
+	/// Convert the xcm into the Message which will be executed
+	/// on Ethereum Gateway contract, we expect an input of the form:
+	/// # WithdrawAsset(WETH)
+	/// # PayFees(WETH)
+	/// # ReserveAssetDeposited(PNA) | WithdrawAsset(ENA)
+	/// # AliasOrigin(Origin)
+	/// # DepositAsset(PNA|ENA)
+	/// # Transact() ---Optional
+	/// # SetTopic
+	fn to_ethereum_message(&mut self) -> Result<Message, XcmConverterError> {
+		use XcmConverterError::*;
+
+		// Get fee amount
+		let fee_amount = self.extract_remote_fee()?;
+
+		// Get ENA reserve asset from WithdrawAsset.
+		let mut enas =
+			match_expression!(self.peek(), Ok(WithdrawAsset(reserve_assets)), reserve_assets);
+		if enas.is_some() {
+			let _ = self.next();
+		}
+
+		// Get PNA reserve asset from ReserveAssetDeposited
+		let pnas = match_expression!(
+			self.peek(),
+			Ok(ReserveAssetDeposited(reserve_assets)),
+			reserve_assets
+		);
+		if pnas.is_some() {
+			let _ = self.next();
+		}
+
+		// Try to get ENA again if it is after PNA
+		if enas.is_none() {
+			enas =
+				match_expression!(self.peek(), Ok(WithdrawAsset(reserve_assets)), reserve_assets);
+			if enas.is_some() {
+				let _ = self.next();
+			}
+		}
+		// Check AliasOrigin.
+		let origin_location = match_expression!(self.next()?, AliasOrigin(origin), origin)
+			.ok_or(AliasOriginExpected)?;
+		let origin = AgentIdOf::convert_location(origin_location).ok_or(InvalidOrigin)?;
+
+		let (deposit_assets, beneficiary) = match_expression!(
+			self.next()?,
+			DepositAsset { assets, beneficiary },
+			(assets, beneficiary)
+		)
+		.ok_or(DepositAssetExpected)?;
+
+		// assert that the beneficiary is AccountKey20.
+		let recipient = match_expression!(
+			beneficiary.unpack(),
+			(0, [AccountKey20 { network, key }])
+				if self.network_matches(network),
+			H160(*key)
+		)
+		.ok_or(BeneficiaryResolutionFailed)?;
+
+		// Make sure there are reserved assets.
+		if enas.is_none() && pnas.is_none() {
+			return Err(NoReserveAssets)
+		}
+
+		let mut commands: Vec<Command> = Vec::new();
+
+		// ENA transfer commands
+		if let Some(enas) = enas {
+			for ena in enas.clone().inner().iter() {
+				// Check the the deposit asset filter matches what was reserved.
+				if !deposit_assets.matches(ena) {
+					return Err(FilterDoesNotConsumeAllAssets)
+				}
+
+				// only fungible asset is allowed
+				let (token, amount) = match ena {
+					Asset { id: AssetId(inner_location), fun: Fungible(amount) } =>
+						match inner_location.unpack() {
+							(0, [AccountKey20 { network, key }])
+								if self.network_matches(network) =>
+								Some((H160(*key), *amount)),
+							_ => None,
+						},
+					_ => None,
+				}
+				.ok_or(AssetResolutionFailed)?;
+
+				// transfer amount must be greater than 0.
+				ensure!(amount > 0, ZeroAssetTransfer);
+
+				commands.push(Command::UnlockNativeToken { token, recipient, amount });
+			}
+		}
+
+		// PNA transfer commands
+		if let Some(pnas) = pnas {
```

Comment:
@yrong please factor this PNA block into a separate function, for ease of reviewing.

Reply:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/65be4090aa811befe67e4d8ec3c384505d907787